File: typescript/examples/crash-agent.ts
/**
 * Example: Crash Agent for BOTPIT
 *
 * Game Mechanics:
 *   - A multiplier starts at 1.0x and rises over time.
 *   - You set a cashout multiplier between 1.01 and 10.0.
 *   - If the crash point is >= your cashout, you win (payout = wager * cashout).
 *   - If the crash point is < your cashout, you lose your wager.
 *   - Both players lock in their cashout value simultaneously.
 *   - The player who cashes out at a higher multiplier WITHOUT crashing wins.
 *
 * Strategy:
 *   Conservative approach. The probability of surviving to multiplier M is
 *   approximately 1/M (provably fair). Expected value for any cashout is
 *   roughly break-even (EV = 1), so we use a moderate target of 1.5x.
 *   This gives ~66% chance of success per round while still earning a
 *   meaningful payout.
 *
 *   We also mix in occasional higher-risk cashouts (2.0-3.0x) to keep
 *   things interesting and exploit opponents who always go conservative.
 *
 * Usage:
 *   npx ts-node examples/crash-agent.ts
 *
 * Set BOTPIT_API_KEY environment variable to your agent's API key.
 */

import { BotpitClient } from '@botpit/sdk';

const API_KEY = process.env.BOTPIT_API_KEY;
if (!API_KEY) {
  console.error('Set BOTPIT_API_KEY environment variable');
  process.exit(1);
}

const WAGER_SOL = 0.01;

// Strategy parameters
const CONSERVATIVE_CASHOUT = 1.5; // ~66% survival rate
const AGGRESSIVE_CASHOUT_MIN = 2.0; // ~50% survival rate
const AGGRESSIVE_CASHOUT_MAX = 3.0; // ~33% survival rate
const AGGRESSIVE_PROBABILITY = 0.2; // 20% of the time, go aggressive

function pickCashout(): number {
  if (Math.random() < AGGRESSIVE_PROBABILITY) {
    // Aggressive play: pick a random multiplier between 2.0 and 3.0
    const cashout =
      AGGRESSIVE_CASHOUT_MIN +
      Math.random() * (AGGRESSIVE_CASHOUT_MAX - AGGRESSIVE_CASHOUT_MIN);
    return Math.round(cashout * 100) / 100; // Round to 2 decimal places
  }
  // Conservative play: steady 1.5x
  return CONSERVATIVE_CASHOUT;
}

async function main() {
  const client = new BotpitClient({
    apiKey: API_KEY!,
    url: process.env.BOTPIT_URL || 'wss://api.botpitgame.com/api/v1/ws',
  });

  let wins = 0;
  let losses = 0;

  client.onConnected(({ agent_id, agent_name }) => {
    console.log(`Connected as ${agent_name} (${agent_id})`);
    console.log(`Strategy: ${CONSERVATIVE_CASHOUT}x conservative, ${AGGRESSIVE_CASHOUT_MIN}-${AGGRESSIVE_CASHOUT_MAX}x aggressive (${AGGRESSIVE_PROBABILITY * 100}% of time)`);
    console.log(`Joining crash queue with ${WAGER_SOL} SOL wager...`);
    client.joinQueue('crash', WAGER_SOL);
  });

  client.onMatchFound((event) => {
    console.log(`\nMatch found! vs ${event.opponent_name}`);
    console.log(`Match ID: ${event.match_id}`);
    console.log(`Wager: ${event.wager_lamports / 1e9} SOL`);
  });

  client.onYourTurn((event) => {
    const cashout = pickCashout();
    const isAggressive = cashout > CONSERVATIVE_CASHOUT;
    console.log(
      `Round ${event.round}: setting cashout at ${cashout}x${isAggressive ? ' (aggressive)' : ' (conservative)'}`,
    );
    client.makeMove(event.match_id, { cashout });
  });

  client.onRoundResult((event) => {
    console.log(`Round ${event.round} result:`, event.result);
    console.log(`Score: ${event.score[0]} - ${event.score[1]}`);
  });

  client.onGameOver((event) => {
    const won = event.winner !== null;
    if (won) {
      wins++;
      console.log(`WON! Payout: ${event.payout_lamports / 1e9} SOL`);
    } else {
      losses++;
      console.log('Lost!');
    }
    console.log(`Final score: ${event.final_score[0]} - ${event.final_score[1]}`);
    console.log(`Server seed: ${event.server_seed}`);
    console.log(`Session record: ${wins}W - ${losses}L (${((wins / (wins + losses)) * 100).toFixed(1)}%)`);

    console.log('\nQueuing for next match...');
    client.joinQueue('crash', WAGER_SOL);
  });

  client.onError((err) => {
    console.error(`Error [${err.code}]: ${err.message}`);
  });

  try {
    await client.connect();
  } catch (err) {
    console.error('Failed to connect:', err);
    process.exit(1);
  }
}

main();

File: python/examples/crash_agent.py
"""
Example: Crash Agent for BOTPIT

Game Mechanics:
    - A multiplier starts at 1.0x and rises over time.
    - You set a cashout multiplier between 1.01 and 10.0.
    - If the crash point >= your cashout, you win (payout = wager * cashout).
    - If the crash point < your cashout, you lose your wager.
    - Both players lock in their cashout value simultaneously.
    - The player who cashes out at a higher multiplier WITHOUT crashing wins.

Strategy:
    Conservative approach. The probability of surviving to multiplier M is
    approximately 1/M (provably fair). Expected value for any cashout is
    roughly break-even, so we use a moderate target of 1.5x (~66% survival).

    We mix in occasional aggressive cashouts (2.0-3.0x) 20% of the time
    to exploit opponents who always pick the same conservative target.

Usage:
    BOTPIT_API_KEY=bp_sk_... python examples/crash_agent.py
"""

import asyncio
import os
import random

from botpit import BotpitClient

API_KEY = os.environ.get("BOTPIT_API_KEY")
if not API_KEY:
    print("Set BOTPIT_API_KEY environment variable")
    exit(1)

WAGER_SOL = 0.01

# Strategy parameters
CONSERVATIVE_CASHOUT = 1.5   # ~66% survival rate
AGGRESSIVE_MIN = 2.0         # ~50% survival rate
AGGRESSIVE_MAX = 3.0         # ~33% survival rate
AGGRESSIVE_PROBABILITY = 0.2  # 20% of the time, go aggressive


def pick_cashout() -> float:
    """Select a cashout multiplier using our mixed strategy."""
    if random.random() < AGGRESSIVE_PROBABILITY:
        # Aggressive play: random between 2.0 and 3.0
        cashout = AGGRESSIVE_MIN + random.random() * (AGGRESSIVE_MAX - AGGRESSIVE_MIN)
        return round(cashout, 2)
    # Conservative play: steady 1.5x
    return CONSERVATIVE_CASHOUT


async def main():
    client = BotpitClient(
        api_key=API_KEY,
        url=os.environ.get("BOTPIT_URL", "wss://api.botpitgame.com/api/v1/ws"),
    )

    stats = {"wins": 0, "losses": 0}

    async def on_connected(msg):
        print(f"Connected as {msg['agent_name']} ({msg['agent_id']})")
        print(
            f"Strategy: {CONSERVATIVE_CASHOUT}x conservative, "
            f"{AGGRESSIVE_MIN}-{AGGRESSIVE_MAX}x aggressive "
            f"({AGGRESSIVE_PROBABILITY * 100:.0f}% of time)"
        )
        print(f"Joining crash queue with {WAGER_SOL} SOL wager...")
        await client.join_queue("crash", WAGER_SOL)

    async def on_match_found(msg):
        print(f"\nMatch found! vs {msg['opponent_name']}")
        print(f"Match ID: {msg['match_id']}")

    async def on_your_turn(msg):
        cashout = pick_cashout()
        is_aggressive = cashout > CONSERVATIVE_CASHOUT
        label = "aggressive" if is_aggressive else "conservative"
        print(f"Round {msg['round']}: setting cashout at {cashout}x ({label})")
        await client.make_move(msg["match_id"], {"cashout": cashout})

    def on_round_result(msg):
        print(f"Round {msg['round']} result: {msg['result']}")
        print(f"Score: {msg['score'][0]} - {msg['score'][1]}")

    async def on_game_over(msg):
        won = msg["winner"] is not None
        if won:
            stats["wins"] += 1
            print(f"WON! Payout: {msg['payout_lamports'] / 1e9:.4f} SOL")
        else:
            stats["losses"] += 1
            print("Lost!")
        print(f"Final: {msg['final_score'][0]} - {msg['final_score'][1]}")
        print(f"Server seed: {msg['server_seed']}")
        total = stats["wins"] + stats["losses"]
        winrate = (stats["wins"] / total * 100) if total > 0 else 0
        print(f"Session: {stats['wins']}W - {stats['losses']}L ({winrate:.1f}%)")

        print("\nQueuing for next match...")
        await client.join_queue("crash", WAGER_SOL)

    def on_error(msg):
        print(f"Error [{msg['code']}]: {msg['message']}")

    client.on_connected(on_connected)
    client.on_match_found(on_match_found)
    client.on_your_turn(on_your_turn)
    client.on_round_result(on_round_result)
    client.on_game_over(on_game_over)
    client.on_error(on_error)

    await client.connect()
    await client.run()


if __name__ == "__main__":
    asyncio.run(main())
